velox/serializers/PrestoSerializer.cpp

```diff
@@ -949,6 +990,24 @@ void VectorStream::append(folly::Range<const IntervalDayTime*> values) {
   }
 }
 
+static const __int128_t kInt128SerializeMask =
+    (static_cast<__int128_t>(1) << 127);
+
+template <>
+inline void VectorStream::append(
+    folly::Range<const UnscaledLongDecimal*> values) {
+  for (auto& value : values) {
+    __int128_t val = value.unscaledValue();
+    // Presto Java UnscaledDecimal128 representation uses signed magnitude
+    // representation. Only negative values differ in this representation.
+    if (val < 0) {
+      val *= -1;
```

Comment:
It looks like the value range of Presto is -(2^(n-1) - 1) to 2^(n-1) - 1 but Velox is -2^(n-1) to 2^(n - 1) - 1. So they don't match? I am wondering how we handle the serialization value of  -2^(n-1)? Shall we take different serialization/deserialization approach that Presto? I assume we don't have the mix the cluster setup with both java and c++ workers? Thanks!

Reply:
@xiaoxmeng good catch. I think the Presto serialization logic should throw an overflow error in this case.

Reply:
For DECIMALS the limits are (10^38 - 1) to -(10^38 - 1) which are at max 127 bits long. So we don't have to expect -2^(n-1) from PRESTO.

```
presto:kart_test> select (cast('170141183460469231731687303715884105727' as decimal(38,0)));
Query 20220902_174639_00021_dg3ps failed: Cannot cast VARCHAR '170141183460469231731687303715884105727' to DECIMAL(38, 0). Value too large.

presto:kart_test> select (cast('-170141183460469231731687303715884105728' as decimal(38,0)));
Query 20220902_193139_00028_dg3ps failed: Cannot cast VARCHAR '-170141183460469231731687303715884105728' to DECIMAL(38, 0). Value too large.

```